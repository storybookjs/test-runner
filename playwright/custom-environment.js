const { setupPage } = require('../dist');

const PlaywrightEnvironment = require('jest-playwright-preset/lib/PlaywrightEnvironment').default;

class CustomEnvironment extends PlaywrightEnvironment {
  async setup() {
    await super.setup();
    await setupPage(this.global.page, this.global.context);
  }

  async teardown() {
    await super.teardown();
  }

  async handleTestEvent(event) {
    await super.handleTestEvent(event);
  }
}

module.exports = CustomEnvironment;
