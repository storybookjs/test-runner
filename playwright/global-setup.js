// global-setup.mjs
import { globalSetup as playwrightGlobalSetup } from 'jest-playwright-preset';

export default async function globalSetup(globalConfig) {
  return playwrightGlobalSetup(globalConfig);
}
