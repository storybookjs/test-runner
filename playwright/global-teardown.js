// global-teardown.mjs
import { globalTeardown as playwrightGlobalTeardown } from 'jest-playwright-preset';

export default async function globalTeardown(globalConfig) {
  // Your global teardown
  await playwrightGlobalTeardown(globalConfig);
}
