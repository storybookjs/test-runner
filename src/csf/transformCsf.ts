/* eslint-disable no-underscore-dangle */
import { toId, storyNameFromExport, combineTags } from '@storybook/csf';
import { loadCsf } from 'storybook/internal/csf-tools';
import * as t from '@babel/types';
import generate from '@babel/generator';
import dedent from 'ts-dedent';

import { getTagOptions } from '../util/getTagOptions';

export interface TestContext {
  storyExport?: t.Identifier;
  name: t.Literal;
  title: t.Literal;
  id: t.Literal;
}
type TemplateResult = t.Statement | t.Statement[];
type FilePrefixer = () => TemplateResult;
export type TestPrefixer = (context: TestContext) => TemplateResult;

export interface TransformOptions {
  clearBody?: boolean;
  beforeEachPrefixer?: FilePrefixer;
  testPrefixer: TestPrefixer;
  insertTestIfEmpty?: boolean;
  makeTitle?: (userTitle: string) => string;
  includeTags?: string[];
  excludeTags?: string[];
  skipTags?: string[];
}

export const prefixFunction = ({
  key,
  testPrefixer,
  title,
  id,
}: {
  key: string;
  title: string;
  testPrefixer: TestPrefixer;
  id?: string;
}) => {
  const name = storyNameFromExport(key);
  const context: TestContext = {
    storyExport: t.identifier(key),
    name: t.stringLiteral(name), // FIXME .name annotation
    title: t.stringLiteral(title),
    id: t.stringLiteral(toId(id ?? title, name)),
  };

  const result = makeArray(testPrefixer(context));
  const stmt = result[1] as t.ExpressionStatement;
  return stmt.expression;
};

const makePlayTest = ({
  key,
  metaOrStoryPlay,
  title,
  id,
  testPrefix,
  shouldSkip,
}: {
  key: string;
  title: string;
  id?: string;
  metaOrStoryPlay?: boolean;
  testPrefix: TestPrefixer;
  shouldSkip?: boolean;
}): t.ExpressionStatement[] => {
  return [
    t.expressionStatement(
      t.callExpression(shouldSkip ? t.identifier('it.skip') : t.identifier('it'), [
        t.stringLiteral(metaOrStoryPlay ? 'play-test' : 'smoke-test'),
        prefixFunction({ key, title, testPrefixer: testPrefix, id }),
      ])
    ),
  ];
};

const makeDescribe = (
  key: string,
  tests: t.Statement[],
  beforeEachBlock?: t.ExpressionStatement
): t.ExpressionStatement => {
  const blockStatements = beforeEachBlock ? [beforeEachBlock, ...tests] : tests;
  return t.expressionStatement(
    t.callExpression(t.identifier('describe'), [
      t.stringLiteral(key),
      t.arrowFunctionExpression([], t.blockStatement(blockStatements)),
    ])
  );
};

const makeBeforeEach = (beforeEachPrefixer: FilePrefixer) => {
  const stmt = beforeEachPrefixer() as t.ExpressionStatement;

  return t.expressionStatement(t.callExpression(t.identifier('beforeEach'), [stmt.expression]));
};

const makeArray = (templateResult: TemplateResult) =>
  Array.isArray(templateResult) ? templateResult : [templateResult];

export const transformCsf = (
  code: string,
  {
    clearBody = false,
    testPrefixer,
    beforeEachPrefixer,
    insertTestIfEmpty,
    makeTitle,
    previewAnnotations = { tags: [] },
  }: TransformOptions & { previewAnnotations?: Record<string, any> }
) => {
  const { includeTags, excludeTags, skipTags } = getTagOptions();

  const csf = loadCsf(code, { makeTitle: makeTitle ?? ((userTitle: string) => userTitle) });
  csf.parse();

  const storyExports = Object.keys(csf._stories);
  const title = csf.meta?.title;

  const storyAnnotations = storyExports.reduce<Record<string, { play?: t.Node; tags?: string[] }>>(
    (acc, key) => {
      const annotations = csf._storyAnnotations[key];
      acc[key] = {};
      if (annotations?.play) {
        // @ts-expect-error type mismatch â€“ check later
        acc[key].play = annotations.play;
      }

      acc[key].tags = combineTags(
        'test',
        'dev',
        ...previewAnnotations.tags,
        ...(csf.meta?.tags || []),
        ...(csf._stories[key].tags || [])
      );

      return acc;
    },
    {}
  );

  const allTests = storyExports
    .filter((key) => {
      // If includeTags is passed, check if the story has any of them - else include by default
      const isIncluded =
        includeTags.length === 0 ||
        includeTags.some((tag) => storyAnnotations[key].tags?.includes(tag));

      // If excludeTags is passed, check if the story does not have any of them
      const isNotExcluded = excludeTags.every((tag) => !storyAnnotations[key].tags?.includes(tag));

      return isIncluded && isNotExcluded;
    })
    .map((key: string) => {
      let tests: t.Statement[] = [];
      const shouldSkip = skipTags.some((tag) => storyAnnotations[key].tags?.includes(tag));
      if (title) {
        tests = [
          ...tests,
          ...makePlayTest({
            key,
            title,
            id: csf.meta.id,
            metaOrStoryPlay: !!storyAnnotations[key]?.play,
            testPrefix: testPrefixer,
            shouldSkip,
          }),
        ];
      }

      if (tests.length) {
        return makeDescribe(key, tests);
      }
    })
    .filter(Boolean) as t.Statement[];

  let result = '';

  if (!clearBody) result = `${result}${code}\n`;
  if (allTests.length) {
    const describe = makeDescribe(
      csf.meta?.title as string,
      allTests,
      beforeEachPrefixer ? makeBeforeEach(beforeEachPrefixer) : undefined
    );
    const { code: describeCode } = generate(describe, {});
    result = dedent`
      ${result}
      if (!require.main) {
        ${describeCode}
      }
    `;
  } else if (insertTestIfEmpty) {
    // When there are no tests at all, we skip. The reason is that the file already went through Jest's transformation,
    // so we have to skip the describe to achieve a "excluded test" experience.
    result = `describe.skip('${csf.meta?.title}', () => { it('no-op', () => {}) });`;
  }
  return result;
};
