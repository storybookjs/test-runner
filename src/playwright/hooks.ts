import type { BrowserContext, Page } from 'playwright';
import type { StoryContextForEnhancers } from '@storybook/csf';

export type TestContext = {
  id: string;
  title: string;
  name: string;
  hasFailure?: boolean;
};

export type PrepareContext = {
  page: Page;
  browserContext: BrowserContext;
  testRunnerConfig: TestRunnerConfig;
};

export type TestHook = (page: Page, context: TestContext) => Promise<void>;
export type HttpHeaderSetter = (url: string) => Promise<Record<string, any>>;
export type PrepareSetter = (context: PrepareContext) => Promise<void>;

export interface TestRunnerConfig {
  setup?: () => void;
  /**
   * @deprecated Use `preVisit` instead.
   */
  preRender?: TestHook;
  /**
   * @deprecated Use `postVisit` instead.
   */
  postRender?: TestHook;
  /**
   * Runs before each story is visited. By this point, the story is not rendered in the browser.
   * This is useful for preparing the browser environment such as setting viewport size, etc.
   * @see https://github.com/storybookjs/test-runner#previsit
   */
  preVisit?: TestHook;
  /**
   * Runs after each story is visited. This means the story has finished rendering and running its play function.
   * This is useful for taking screenshots, snapshots, accessibility tests, etc.
   * @see https://github.com/storybookjs/test-runner#postvisit
   */
  postVisit?: TestHook;
  /**
   * Adds http headers to the test-runner's requests. This is useful if you need to set headers such as `Authorization` for your Storybook instance.
   */
  getHttpHeaders?: HttpHeaderSetter;
  /**
   * Overrides the default prepare behavior of the test-runner. Good for customizing the environment before testing, such as authentication etc.
   *
   * If you override the default prepare behavior, even though this is powerful, you will be responsible for properly preparing the browser. Future changes to the default prepare function will not get included in your project, so you will have to keep an eye out for changes in upcoming releases.
   */
  prepare?: PrepareSetter;
  /**
   * Tags to include, exclude, or skip. These tags are defined as annotations in your story or meta.
   * @see https://github.com/storybookjs/test-runner#filtering-tests-experimental
   */
  tags?: {
    include?: string[];
    exclude?: string[];
    skip?: string[];
  };
  /**
   * Defines the log level of the test runner. Browser logs are printed to the console when reporting errors.
   * @default 'info'
   */
  logLevel?: 'info' | 'warn' | 'error' | 'verbose' | 'none';

  /**
   * Defines a custom function to process the error message. Useful to sanitize error messages or to add additional information.
   */
  errorMessageFormatter?: (error: string) => string;
}

export const setPreVisit = (preVisit: TestHook) => {
  globalThis.__sbPreVisit = preVisit;
};

export const setPostVisit = (postVisit: TestHook) => {
  globalThis.__sbPostVisit = postVisit;
};

export const getStoryContext = async (
  page: Page,
  context: TestContext
): Promise<StoryContextForEnhancers> => {
  return page.evaluate(({ storyId }) => globalThis.__getContext(storyId), {
    storyId: context.id,
  });
};

export const waitForPageReady = async (page: Page) => {
  await page.waitForLoadState('domcontentloaded');
  await page.waitForLoadState('load');
  await page.waitForLoadState('networkidle');
  await page.evaluate(() => document.fonts.ready);
};
