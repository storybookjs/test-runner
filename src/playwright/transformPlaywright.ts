import { relative } from 'path';
import template from '@babel/template';
import { userOrAutoTitle } from 'storybook/internal/preview-api';
import dedent from 'ts-dedent';

import { getStorybookMetadata } from '../util';
import { transformCsf } from '../csf/transformCsf';
import type { TestPrefixer } from '../csf/transformCsf';

const coverageErrorMessage = dedent`
  [Test runner] An error occurred when evaluating code coverage:
  The code in this story is not instrumented, which means the coverage setup is likely not correct.
  More info: https://github.com/storybookjs/test-runner#setting-up-code-coverage
`;

export const testPrefixer: TestPrefixer = (context) => {
  return template(
    `
    console.log({ id: %%id%%, title: %%title%%, name: %%name%%, storyExport: %%storyExport%% });
    async () => {
      const testFn = async() => {
        const context = { id: %%id%%, title: %%title%%, name: %%name%% };

        if(globalThis.__sbPreVisit) {
          await globalThis.__sbPreVisit(page, context);
        }

        let result;
        try {
          result = await page.evaluate(({ id, hasPlayFn }) => __test(id, hasPlayFn), {
            id: %%id%%,
          });
        } catch(err) {
          if(err.toString().includes('Execution context was destroyed')) {
            // Retry the test, possible Vite dep optimization flake
            throw err;
          } else {
            await globalThis.__sbPostVisit(page, {...context, hasFailure: true });
            throw err;
          }
        }
  
        if(globalThis.__sbPostVisit) {
          await globalThis.__sbPostVisit(page, context);
        }

        if(globalThis.__sbCollectCoverage) {
        const isCoverageSetupCorrectly = await page.evaluate(() => '__coverage__' in window);
          if (!isCoverageSetupCorrectly) {
            throw new Error(\`${coverageErrorMessage}\`);
          }

          await jestPlaywright.saveCoverage(page);
        }


        return result;
      };

      try {
        await testFn();
      } catch(err) {
        if(err.toString().includes('Execution context was destroyed')) {
          console.log(\`An error occurred in the following story, most likely because of a navigation: "\${%%title%%}/\${%%name%%}". Retrying...\`);
          await jestPlaywright.resetPage();
          await globalThis.__sbSetupPage(globalThis.page, globalThis.context);
          await testFn();
        } else {
          throw err;
        }
      }
    }
  `,
    {
      plugins: ['jsx'],
    }
  )({ ...context });
};

const makeTitleFactory = (filename: string) => {
  const { workingDir, normalizedStoriesEntries } = getStorybookMetadata();
  const filePath = `./${relative(workingDir, filename)}`;

  return (userTitle: string) =>
    userOrAutoTitle(filePath, normalizedStoriesEntries, userTitle) as string;
};

export const transformPlaywright = (src: string, filename: string) => {
  const tags = process.env.STORYBOOK_PREVIEW_TAGS?.split(',') ?? [];
  const transformOptions = {
    testPrefixer,
    insertTestIfEmpty: true,
    clearBody: true,
    makeTitle: makeTitleFactory(filename),
    previewAnnotations: { tags },
  };

  const result = transformCsf(src, transformOptions);
  return result;
};
