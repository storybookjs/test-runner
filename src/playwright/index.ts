import { transformSync as swcTransform } from '@swc/core';
import { transformPlaywright } from './transformPlaywright';

export const processAsync = async (src: string, filename: string) => {
  const csfTest = await transformPlaywright(src, filename);

  const result = swcTransform(csfTest, {
    filename,
    module: {
      type: 'commonjs',
    },
  });
  return result ? result.code : src;
};
