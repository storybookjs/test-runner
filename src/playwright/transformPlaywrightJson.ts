import * as t from '@babel/types';
import generate from '@babel/generator';
import { ComponentTitle, StoryId, StoryName, toId } from '@storybook/csf';

import { testPrefixer } from './transformPlaywright';
import { getTagOptions } from '../util/getTagOptions';

const makeTest = ({
  entry,
  shouldSkip,
  metaOrStoryPlay,
}: {
  entry: V4Entry;
  shouldSkip: boolean;
  metaOrStoryPlay: boolean;
}): t.Statement => {
  const result = testPrefixer({
    name: t.stringLiteral(entry.name),
    title: t.stringLiteral(entry.title),
    id: t.stringLiteral(entry.id),
    // FIXME
    storyExport: t.identifier(entry.id),
  });
  const stmt = (result as Array<t.ExpressionStatement>)[1];
  return t.expressionStatement(
    t.callExpression(shouldSkip ? t.identifier('it.skip') : t.identifier('it'), [
      t.stringLiteral(metaOrStoryPlay ? 'play-test' : 'smoke-test'),
      stmt.expression,
    ])
  );
};

export const makeDescribe = (title: string, stmts: t.Statement[]) => {
  // When there are no tests at all, we skip. The reason is that the file already went through Jest's transformation,
  // so we have to skip the describe to achieve a "excluded test" experience.
  // The code below recreates the following source:
  // describe.skip(`${title}`, () => { it('no-op', () => {}) });
  if (stmts.length === 0) {
    const noOpIt = t.expressionStatement(
      t.callExpression(t.identifier('it'), [
        t.stringLiteral('no-op'),
        t.arrowFunctionExpression([], t.blockStatement([])),
      ])
    );

    return t.expressionStatement(
      t.callExpression(t.memberExpression(t.identifier('describe'), t.identifier('skip')), [
        t.stringLiteral(title),
        t.arrowFunctionExpression([], t.blockStatement([noOpIt])),
      ])
    );
  }

  return t.expressionStatement(
    t.callExpression(t.identifier('describe'), [
      t.stringLiteral(title),
      t.arrowFunctionExpression([], t.blockStatement(stmts)),
    ])
  );
};

type V3Story = Omit<V4Entry, 'type'> & { parameters?: StoryParameters };
export type V3StoriesIndex = {
  v: 3;
  stories: Record<StoryId, V3Story>;
};
type V4Entry = {
  type?: 'story' | 'docs';
  id: StoryId;
  name: StoryName;
  title: ComponentTitle;
  tags?: string[];
};
export type V4Index = {
  v: 4;
  entries: Record<StoryId, V4Entry>;
};

type V5Entry = V4Entry & { tags: string[] };
export type V5Index = {
  v: 5;
  entries: Record<StoryId, V5Entry>;
};

type StoryParameters = {
  __id: StoryId;
  docsOnly?: boolean;
  fileName?: string;
};

export type UnsupportedVersion = { v: number };
const isV3DocsOnly = (stories: V3Story[]) => stories.length === 1 && stories[0].name === 'Page';

function v3TitleMapToV4TitleMap(titleIdToStories: Record<string, V3Story[]>) {
  return Object.fromEntries(
    Object.entries(titleIdToStories).map(([id, stories]) => [
      id,
      stories.map(
        ({ parameters, ...story }) =>
          ({
            type: isV3DocsOnly(stories) ? 'docs' : 'story',
            tags: isV3DocsOnly(stories) ? [] : ['test', 'dev'],
            ...story,
          }) satisfies V4Entry
      ),
    ])
  );
}

/**
 * Storybook 8.0 and below did not automatically tag stories with 'dev'.
 * Therefore Storybook 8.1 and above would not show composed 8.0 stories by default.
 * This function adds the 'dev' tag to all stories in the index to workaround this issue.
 */
function v4TitleMapToV5TitleMap(titleIdToStories: Record<string, V4Entry[]>) {
  return Object.fromEntries(
    Object.entries(titleIdToStories).map(([id, stories]) => [
      id,
      stories.map(
        (story) =>
          ({
            ...story,
            tags: story.tags ? ['test', 'dev', ...story.tags] : ['test', 'dev'],
          }) satisfies V4Entry
      ),
    ])
  );
}

function groupByTitleId<T extends { title: ComponentTitle }>(entries: T[]) {
  return entries.reduce<Record<string, T[]>>((acc, entry) => {
    const titleId = toId(entry.title);
    acc[titleId] = acc[titleId] || [];
    acc[titleId].push(entry);
    return acc;
  }, {});
}

/**
 * Generate one test file per component so that Jest can
 * run them in parallel.
 */
export const transformPlaywrightJson = (index: V3StoriesIndex | V4Index | UnsupportedVersion) => {
  let titleIdToEntries: Record<string, V4Entry[]>;
  if (index.v === 3) {
    const titleIdToStories = groupByTitleId<V3Story>(
      Object.values((index as V3StoriesIndex).stories)
    );
    titleIdToEntries = v3TitleMapToV4TitleMap(titleIdToStories);
  } else if (index.v === 4) {
    // TODO: Once Storybook 8.0 is released, we should only support v4 and higher
    titleIdToEntries = groupByTitleId<V4Entry>(Object.values((index as V4Index).entries));
    titleIdToEntries = v4TitleMapToV5TitleMap(titleIdToEntries);
  } else if (index.v === 5) {
    titleIdToEntries = groupByTitleId<V4Entry>(Object.values((index as V4Index).entries));
  } else {
    throw new Error(`Unsupported version ${index.v}`);
  }

  const { includeTags, excludeTags, skipTags } = getTagOptions();

  const titleIdToTest = Object.entries(titleIdToEntries).reduce<Record<string, string>>(
    (acc, [titleId, entries]) => {
      const stories = entries.filter((s) => s.type !== 'docs');
      if (stories.length) {
        const storyTests = stories
          .filter((story) => {
            // If includeTags is passed, check if the story has any of them - else include by default
            const isIncluded =
              includeTags.length === 0 || includeTags.some((tag) => story.tags?.includes(tag));

            // If excludeTags is passed, check if the story does not have any of them
            const isNotExcluded = excludeTags.every((tag) => !story.tags?.includes(tag));

            return isIncluded && isNotExcluded;
          })
          .map((story) => {
            const shouldSkip = skipTags.some((tag) => story.tags?.includes(tag));

            return makeDescribe(story.name, [
              makeTest({
                entry: story,
                shouldSkip,
                metaOrStoryPlay: story.tags?.includes('play-fn') ?? false,
              }),
            ]);
          });
        const program = t.program([makeDescribe(stories[0].title, storyTests)]);

        const { code } = generate(program, {});

        acc[titleId] = code;
      }
      return acc;
    },
    {}
  );

  return titleIdToTest;
};
