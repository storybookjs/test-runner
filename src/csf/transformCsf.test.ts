import { TestPrefixer, TransformOptions, transformCsf } from './transformCsf';
import { testPrefixer } from '../playwright/transformPlaywright';
import template from '@babel/template';
import { describe, it, expect } from 'vitest';

describe('transformCsf', async () => {
  it('inserts a no-op test if there are no stories', async () => {
    const csfCode = `
      export default {
        title: 'Button',
      };
    `;
    const expectedCode = `describe.skip('Button', () => { it('no-op', () => {}) });`;

    const result = await transformCsf(csfCode, { insertTestIfEmpty: true } as TransformOptions);

    expect(result).toEqual(expectedCode);
  });

  it('returns empty result if there are no stories', async () => {
    const csfCode = `
      export default {
        title: 'Button',
      };
    `;

    const result = await transformCsf(csfCode, { testPrefixer });

    expect(result).toMatchSnapshot();
  });

  it('calls the testPrefixer function for each test', async () => {
    const csfCode = `
      export default {
        title: 'Button',
        parameters: {
          play: {
            steps: [
              { id: 'step1', action: 'click', target: 'button' },
              { id: 'step2', action: 'click', target: 'button' },
            ],
          },
        },
      };
      export const Primary = () => '<button>Primary</button>';
    `;

    const result = await transformCsf(csfCode, { testPrefixer });

    expect(result).toMatchSnapshot();
  });

  it('calls the beforeEachPrefixer function once', async () => {
    const csfCode = `
      export default {
        title: 'Button',
        parameters: {
          play: {
            steps: [
              { id: 'step1', action: 'click', target: 'button' },
              { id: 'step2', action: 'click', target: 'button' },
            ],
          },
        },
      };
      export const Primary = () => '<button>Primary</button>';
    `;
    const result = await transformCsf(csfCode, { testPrefixer, beforeEachPrefixer: undefined });

    expect(result).toMatchSnapshot();
  });

  it('clears the body if clearBody option is true', async () => {
    const csfCode = `
      export default {
        title: 'Button',
        parameters: {
          play: {
            steps: [
              { id: 'step1', action: 'click', target: 'button' },
              { id: 'step2', action: 'click', target: 'button' },
            ],
          },
        },
      };
      export const Primary = () => '<button>Primary</button>';
    `;

    const result = await transformCsf(csfCode, { testPrefixer, clearBody: true });

    expect(result).toMatchSnapshot();
  });

  it('executes beforeEach code before each test', async () => {
    const code = `
    export default {
        title: 'Button',
        parameters: {
          play: {
            steps: [
              { id: 'step1', action: 'click', target: 'button' },
              { id: 'step2', action: 'click', target: 'button' },
            ],
          },
        },
      };
      export const Primary = () => '<button>Primary</button>';
    `;
    const beforeEachPrefixer = () => {
      const logStatement = template.expression`console.log("beforeEach called")`;
      const beforeEachBlock = template.statement`beforeEach(() => { ${logStatement()} })`;
      return beforeEachBlock();
    };
    const testPrefixer = template(`
      console.log({ id: %%id%%, title: %%title%%, name: %%name%%, storyExport: %%storyExport%% });
      async () => {}`) as unknown as TestPrefixer;

    const result = await transformCsf(code, {
      beforeEachPrefixer,
      testPrefixer,
    } as TransformOptions);

    expect(result).toMatchSnapshot();
  });
});
