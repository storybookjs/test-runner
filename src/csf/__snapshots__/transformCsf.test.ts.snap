// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformCsf calls the beforeEachPrefixer function once 1`] = `
"
      export default {
        title: 'Button',
        parameters: {
          play: {
            steps: [
              { id: 'step1', action: 'click', target: 'button' },
              { id: 'step2', action: 'click', target: 'button' },
            ],
          },
        },
      };
      export const Primary = () => '<button>Primary</button>';
    

if (!require.main) {
  describe("Button", () => {
    describe("Primary", () => {
      it("smoke-test", async () => {
        const testFn = async () => {
          const context = {
            id: "button--primary",
            title: "Button",
            name: "Primary"
          };
          if (globalThis.__sbPreVisit) {
            await globalThis.__sbPreVisit(page, context);
          }
          let result;
          try {
            result = await page.evaluate(({
              id,
              hasPlayFn
            }) => __test(id, hasPlayFn), {
              id: "button--primary"
            });
          } catch (err) {
            if (err.toString().includes('Execution context was destroyed')) {
              throw err;
            } else {
              if (globalThis.__sbPostVisit) {
                await globalThis.__sbPostVisit(page, {
                  ...context,
                  hasFailure: true
                });
              }
              throw err;
            }
          }
          if (globalThis.__sbPostVisit) {
            await globalThis.__sbPostVisit(page, context);
          }
          if (globalThis.__sbCollectCoverage) {
            const isCoverageSetupCorrectly = await page.evaluate(() => '__coverage__' in window);
            if (!isCoverageSetupCorrectly) {
              throw new Error(\`[Test runner] An error occurred when evaluating code coverage:
  The code in this story is not instrumented, which means the coverage setup is likely not correct.
  More info: https://github.com/storybookjs/test-runner#setting-up-code-coverage\`);
            }
            await jestPlaywright.saveCoverage(page);
          }
          return result;
        };
        try {
          await testFn();
        } catch (err) {
          if (err.toString().includes('Execution context was destroyed')) {
            console.log(\`An error occurred in the following story, most likely because of a navigation: "\${"Button"}/\${"Primary"}". Retrying...\`);
            await jestPlaywright.resetPage();
            await globalThis.__sbSetupPage(globalThis.page, globalThis.context);
            await testFn();
          } else {
            throw err;
          }
        }
      });
    });
  });
}"
`;

exports[`transformCsf calls the testPrefixer function for each test 1`] = `
"
      export default {
        title: 'Button',
        parameters: {
          play: {
            steps: [
              { id: 'step1', action: 'click', target: 'button' },
              { id: 'step2', action: 'click', target: 'button' },
            ],
          },
        },
      };
      export const Primary = () => '<button>Primary</button>';
    

if (!require.main) {
  describe("Button", () => {
    describe("Primary", () => {
      it("smoke-test", async () => {
        const testFn = async () => {
          const context = {
            id: "button--primary",
            title: "Button",
            name: "Primary"
          };
          if (globalThis.__sbPreVisit) {
            await globalThis.__sbPreVisit(page, context);
          }
          let result;
          try {
            result = await page.evaluate(({
              id,
              hasPlayFn
            }) => __test(id, hasPlayFn), {
              id: "button--primary"
            });
          } catch (err) {
            if (err.toString().includes('Execution context was destroyed')) {
              throw err;
            } else {
              if (globalThis.__sbPostVisit) {
                await globalThis.__sbPostVisit(page, {
                  ...context,
                  hasFailure: true
                });
              }
              throw err;
            }
          }
          if (globalThis.__sbPostVisit) {
            await globalThis.__sbPostVisit(page, context);
          }
          if (globalThis.__sbCollectCoverage) {
            const isCoverageSetupCorrectly = await page.evaluate(() => '__coverage__' in window);
            if (!isCoverageSetupCorrectly) {
              throw new Error(\`[Test runner] An error occurred when evaluating code coverage:
  The code in this story is not instrumented, which means the coverage setup is likely not correct.
  More info: https://github.com/storybookjs/test-runner#setting-up-code-coverage\`);
            }
            await jestPlaywright.saveCoverage(page);
          }
          return result;
        };
        try {
          await testFn();
        } catch (err) {
          if (err.toString().includes('Execution context was destroyed')) {
            console.log(\`An error occurred in the following story, most likely because of a navigation: "\${"Button"}/\${"Primary"}". Retrying...\`);
            await jestPlaywright.resetPage();
            await globalThis.__sbSetupPage(globalThis.page, globalThis.context);
            await testFn();
          } else {
            throw err;
          }
        }
      });
    });
  });
}"
`;

exports[`transformCsf clears the body if clearBody option is true 1`] = `
"
if (!require.main) {
  describe("Button", () => {
    describe("Primary", () => {
      it("smoke-test", async () => {
        const testFn = async () => {
          const context = {
            id: "button--primary",
            title: "Button",
            name: "Primary"
          };
          if (globalThis.__sbPreVisit) {
            await globalThis.__sbPreVisit(page, context);
          }
          let result;
          try {
            result = await page.evaluate(({
              id,
              hasPlayFn
            }) => __test(id, hasPlayFn), {
              id: "button--primary"
            });
          } catch (err) {
            if (err.toString().includes('Execution context was destroyed')) {
              throw err;
            } else {
              if (globalThis.__sbPostVisit) {
                await globalThis.__sbPostVisit(page, {
                  ...context,
                  hasFailure: true
                });
              }
              throw err;
            }
          }
          if (globalThis.__sbPostVisit) {
            await globalThis.__sbPostVisit(page, context);
          }
          if (globalThis.__sbCollectCoverage) {
            const isCoverageSetupCorrectly = await page.evaluate(() => '__coverage__' in window);
            if (!isCoverageSetupCorrectly) {
              throw new Error(\`[Test runner] An error occurred when evaluating code coverage:
  The code in this story is not instrumented, which means the coverage setup is likely not correct.
  More info: https://github.com/storybookjs/test-runner#setting-up-code-coverage\`);
            }
            await jestPlaywright.saveCoverage(page);
          }
          return result;
        };
        try {
          await testFn();
        } catch (err) {
          if (err.toString().includes('Execution context was destroyed')) {
            console.log(\`An error occurred in the following story, most likely because of a navigation: "\${"Button"}/\${"Primary"}". Retrying...\`);
            await jestPlaywright.resetPage();
            await globalThis.__sbSetupPage(globalThis.page, globalThis.context);
            await testFn();
          } else {
            throw err;
          }
        }
      });
    });
  });
}"
`;

exports[`transformCsf executes beforeEach code before each test 1`] = `
"
    export default {
        title: 'Button',
        parameters: {
          play: {
            steps: [
              { id: 'step1', action: 'click', target: 'button' },
              { id: 'step2', action: 'click', target: 'button' },
            ],
          },
        },
      };
      export const Primary = () => '<button>Primary</button>';
    

if (!require.main) {
  describe("Button", () => {
    beforeEach(beforeEach(() => {
      console.log("beforeEach called");
    }));
    describe("Primary", () => {
      it("smoke-test", async () => {});
    });
  });
}"
`;

exports[`transformCsf returns empty result if there are no stories 1`] = `
"
      export default {
        title: 'Button',
      };
    
"
`;
